blscurve/bls_backend.nim

```diff
@@ -20,28 +20,36 @@ type BlsBackendKind* = enum
   BLST
   Miracl
 
-const AutoSelectBLST = BLS_FORCE_BACKEND == "auto" and (
-  defined(arm64) or defined(arm) or
-  defined(amd64) or defined(i386)
-)
-# Theoretically the BLST library has a fallback for any platform
-# but it is missing https://github.com/supranational/blst/issues/46
-
-when (BLS_FORCE_BACKEND == "blst" or AutoSelectBLST) and (
-  gorgeEx(getEnv("CC", "gcc") & " -march=native -dM -E -x c /dev/null | grep -q SSSE3").exitCode == 0
-  ):
-  # BLST supports: x86 and ARM 32 and 64 bits
-  # and has optimized SHA256 routines for x86_64 CPU with SSE3
-  # It also assumes that all ARM CPUs are Neon instructions capable for SHA256
-  const BLS_BACKEND* = BLST
-elif BLS_FORCE_BACKEND == "blst" or AutoSelectBLST:
-  # CPU doesn't support SSE3 which is used in optimized SHA256
-  # On ARM, BLST_PORTABLE will prevent use builtin SHA256
-  # which is unsupported by Raspberry Pi, detection via (__ARM_FEATURE_CRYPTO)
+const UseBLST = BLS_FORCE_BACKEND == "auto" or BLS_FORCE_BACKEND == "blst"
+const OnX86 = defined(i386) or defined(amd64)
+const OnARM = defined(arm) or defined(arm64)
+
+when UseBLST:
+  when OnX86:
+    # BLST defaults to SSE3 for SHA256 (Pentium 4, 2004)
+    # And autodetects MULX and ADCX/ADOX for bigints (Intel Broadwell 2015, AMD Ryzen 2017)
+    # It is set either via -march=native on a proper CPU
+    # or -madx
+    const useSSE3 = gorgeEx(getEnv("CC", "gcc") & " -march=native -dM -E -x c /dev/null | grep -q SSSE3").exitCode == 0
```

Comment:
we have some mode that runs without `-march=native` afair to create portable binaries.. @stefantalpalaru ?

Reply:
Our "no -march-native" mode comes with "-msse3": https://github.com/status-im/nimbus-eth2/blob/8ad3b3f551000c88047ec687d55cc02295539b0f/config.nims#L43

Reply:
my point was more that this check will use `march=native` always which seems.. wrong?

Reply:
ie it won't work for portable and cross compiles when imported

Reply:
To clarify, those are the scenarios:

1. The CPU we build on supports SSE3, this is detected by the gorgeEx procedure, if so BLST is build as default i.e. requiring SSE3.
2. The CPU we build on doesn't support SSE3, this is detected by gorgeEx, if so we pass `{.passC: "-D__BLST_PORTABLE__".}` This is always the case for cross-compilation from say ARM to x86-64 and also the case for one of our Jenkins runner.

We can always pass `{.passC: "-D__BLST_PORTABLE__".}` in Nimbus. Detecting SSE3 support on the CPU means we do nothing, in particular we don't pass `-msse3` so we don't pollute GCC arch options.

Lastly, I didn't change the logic, just reorganized the code.

Reply:
We don't support non-SSE3 AMD64 CPUs, so the only remaining problem is if the CI host where we build our Nimbus-eth2 binaries wouldn't support SSE3, switching off SSE3 support for everyone - highly unlikely.

A similar problem would be if the CI host supports ADX and enables it due to automagic detection. That we need to explicitly disable in https://github.com/status-im/nimbus-eth2/blob/c4d9d5f64e32348dd7ec1dc49572f4e97149d8d9/.github/workflows/release.yml#L20

Any C define for that, @mratsim?

Reply:
I looked into it and we can't easily disable ADX without also disabling SSE3 (`__BLST_PORTABLE__` disables all accelerations).